pkg/bitcoin/electrum/electrum_integration_test.go

```diff
@@ -0,0 +1,354 @@
+//go:build integration
+// +build integration
+
+package electrum
+
+import (
+	"context"
+	"encoding/hex"
+	"fmt"
+	"math"
+	"testing"
+	"time"
+
+	"github.com/go-test/deep"
+	"golang.org/x/exp/slices"
+
+	"github.com/keep-network/keep-core/pkg/bitcoin"
+)
+
+// TODO: Include integration test in the CI.
+// To run the tests execute `go test -v -tags=integration ./...`
```

Comment:
@nkuba I ran the integration tests a couple of time, but still getting quite a few errors:
```
tslabon@MacBook-Pro-Tomasz ~/projekt/keep-core $% go test -v -tags=integration ./pkg/bitcoin/electrum
=== RUN   TestGetTransaction_Integration
=== RUN   TestGetTransaction_Integration/electrs-esplora_tcp
=== RUN   TestGetTransaction_Integration/electrs-esplora_ssl
=== RUN   TestGetTransaction_Integration/electrumx_ssl
=== RUN   TestGetTransaction_Integration/fulcrum_ssl
--- PASS: TestGetTransaction_Integration (2.91s)
    --- PASS: TestGetTransaction_Integration/electrs-esplora_tcp (0.46s)
    --- PASS: TestGetTransaction_Integration/electrs-esplora_ssl (0.64s)
    --- PASS: TestGetTransaction_Integration/electrumx_ssl (0.97s)
    --- PASS: TestGetTransaction_Integration/fulcrum_ssl (0.84s)
=== RUN   TestGetTransaction_Negative_Integration
=== RUN   TestGetTransaction_Negative_Integration/electrs-esplora_tcp
    electrum_integration_test.go:118: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [missing transaction]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [missing transaction]]]
=== RUN   TestGetTransaction_Negative_Integration/electrs-esplora_ssl
    electrum_integration_test.go:118: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [missing transaction]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [missing transaction]]]
=== RUN   TestGetTransaction_Negative_Integration/electrumx_ssl
    electrum_integration_test.go:118: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
=== RUN   TestGetTransaction_Negative_Integration/fulcrum_ssl
    electrum_integration_test.go:118: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
--- FAIL: TestGetTransaction_Negative_Integration (6.08s)
    --- FAIL: TestGetTransaction_Negative_Integration/electrs-esplora_tcp (1.30s)
    --- FAIL: TestGetTransaction_Negative_Integration/electrs-esplora_ssl (1.49s)
    --- FAIL: TestGetTransaction_Negative_Integration/electrumx_ssl (1.67s)
    --- FAIL: TestGetTransaction_Negative_Integration/fulcrum_ssl (1.61s)
=== RUN   TestGetTransactionConfirmations_Integration
=== RUN   TestGetTransactionConfirmations_Integration/electrs-esplora_tcp
=== RUN   TestGetTransactionConfirmations_Integration/electrs-esplora_ssl
=== RUN   TestGetTransactionConfirmations_Integration/electrumx_ssl
=== RUN   TestGetTransactionConfirmations_Integration/fulcrum_ssl
--- PASS: TestGetTransactionConfirmations_Integration (4.23s)
    --- PASS: TestGetTransactionConfirmations_Integration/electrs-esplora_tcp (0.81s)
    --- PASS: TestGetTransactionConfirmations_Integration/electrs-esplora_ssl (0.95s)
    --- PASS: TestGetTransactionConfirmations_Integration/electrumx_ssl (1.25s)
    --- PASS: TestGetTransactionConfirmations_Integration/fulcrum_ssl (1.21s)
=== RUN   TestGetTransactionConfirmations_Negative_Integration
=== RUN   TestGetTransactionConfirmations_Negative_Integration/fulcrum_ssl
    electrum_integration_test.go:185: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
=== RUN   TestGetTransactionConfirmations_Negative_Integration/electrs-esplora_tcp
    electrum_integration_test.go:185: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [missing transaction]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [missing transaction]]]
=== RUN   TestGetTransactionConfirmations_Negative_Integration/electrs-esplora_ssl
    electrum_integration_test.go:185: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [missing transaction]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [missing transaction]]]
=== RUN   TestGetTransactionConfirmations_Negative_Integration/electrumx_ssl
    electrum_integration_test.go:185: invalid error
        expected: failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [GetRawTransaction failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
        actual:   failed to get raw transaction with ID [ecc246ac58e682c8edccabb6476bb5482df541847b774085cdb8bfc53165cd34]: [retry timeout [1s] exceeded; most recent error: [request failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
--- FAIL: TestGetTransactionConfirmations_Negative_Integration (6.18s)
    --- FAIL: TestGetTransactionConfirmations_Negative_Integration/fulcrum_ssl (1.58s)
    --- FAIL: TestGetTransactionConfirmations_Negative_Integration/electrs-esplora_tcp (1.31s)
    --- FAIL: TestGetTransactionConfirmations_Negative_Integration/electrs-esplora_ssl (1.48s)
    --- FAIL: TestGetTransactionConfirmations_Negative_Integration/electrumx_ssl (1.81s)
=== RUN   TestGetLatestBlockHeight_Integration
=== RUN   TestGetLatestBlockHeight_Integration/fulcrum_ssl
=== RUN   TestGetLatestBlockHeight_Integration/electrs-esplora_tcp
    electrum_integration_test.go:204: failed to subscribe for headers: [retry timeout [1s] exceeded; most recent error: [request failed: [failed to get the blocks tip height: [request timeout]]]]
=== RUN   TestGetLatestBlockHeight_Integration/electrs-esplora_ssl
=== RUN   TestGetLatestBlockHeight_Integration/electrumx_ssl
--- FAIL: TestGetLatestBlockHeight_Integration (46.47s)
    --- PASS: TestGetLatestBlockHeight_Integration/fulcrum_ssl (0.76s)
    --- FAIL: TestGetLatestBlockHeight_Integration/electrs-esplora_tcp (1.31s)
    --- PASS: TestGetLatestBlockHeight_Integration/electrs-esplora_ssl (43.42s)
    --- PASS: TestGetLatestBlockHeight_Integration/electrumx_ssl (0.98s)
=== RUN   TestGetBlockHeader_Integration
=== RUN   TestGetBlockHeader_Integration/electrs-esplora_tcp
=== RUN   TestGetBlockHeader_Integration/electrs-esplora_ssl
=== RUN   TestGetBlockHeader_Integration/electrumx_ssl
=== RUN   TestGetBlockHeader_Integration/fulcrum_ssl
--- PASS: TestGetBlockHeader_Integration (2.79s)
    --- PASS: TestGetBlockHeader_Integration/electrs-esplora_tcp (0.48s)
    --- PASS: TestGetBlockHeader_Integration/electrs-esplora_ssl (0.67s)
    --- PASS: TestGetBlockHeader_Integration/electrumx_ssl (0.82s)
    --- PASS: TestGetBlockHeader_Integration/fulcrum_ssl (0.82s)
=== RUN   TestGetBlockHeader_Negative_Integration
=== RUN   TestGetBlockHeader_Negative_Integration/electrs-esplora_tcp
    electrum_integration_test.go:288: invalid error
        expected: failed to get block header: [retry timeout [1s] exceeded; most recent error: [GetBlockHeader failed: [missing header]]]
        actual:   failed to get block header: [retry timeout [1s] exceeded; most recent error: [request failed: [missing header]]]
=== RUN   TestGetBlockHeader_Negative_Integration/electrs-esplora_ssl
    electrum_integration_test.go:288: invalid error
        expected: failed to get block header: [retry timeout [1s] exceeded; most recent error: [GetBlockHeader failed: [missing header]]]
        actual:   failed to get block header: [retry timeout [1s] exceeded; most recent error: [request failed: [missing header]]]
=== RUN   TestGetBlockHeader_Negative_Integration/electrumx_ssl
    electrum_integration_test.go:288: invalid error
        expected: failed to get block header: [retry timeout [1s] exceeded; most recent error: [GetBlockHeader failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
        actual:   failed to get block header: [retry timeout [1s] exceeded; most recent error: [request failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
=== RUN   TestGetBlockHeader_Negative_Integration/fulcrum_ssl
    electrum_integration_test.go:288: invalid error
        expected: failed to get block header: [retry timeout [1s] exceeded; most recent error: [GetBlockHeader failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
        actual:   failed to get block header: [retry timeout [1s] exceeded; most recent error: [request failed: [Unmarshal received message failed: json: cannot unmarshal object into Go struct field response.error of type string]]]
--- FAIL: TestGetBlockHeader_Negative_Integration (6.08s)
    --- FAIL: TestGetBlockHeader_Negative_Integration/electrs-esplora_tcp (1.33s)
    --- FAIL: TestGetBlockHeader_Negative_Integration/electrs-esplora_ssl (1.51s)
    --- FAIL: TestGetBlockHeader_Negative_Integration/electrumx_ssl (1.64s)
    --- FAIL: TestGetBlockHeader_Negative_Integration/fulcrum_ssl (1.60s)
=== RUN   TestParseProtocol
=== RUN   TestParseProtocol/lower_case:_tcp
=== RUN   TestParseProtocol/upper_case:_TCP
=== RUN   TestParseProtocol/mixed_case:_Tcp
=== RUN   TestParseProtocol/lower_case:_ssl
=== RUN   TestParseProtocol/upper_case:_SSL
=== RUN   TestParseProtocol/mixed_case:_SsL
=== RUN   TestParseProtocol/invalid:_tls
=== RUN   TestParseProtocol/empty
--- PASS: TestParseProtocol (0.00s)
    --- PASS: TestParseProtocol/lower_case:_tcp (0.00s)
    --- PASS: TestParseProtocol/upper_case:_TCP (0.00s)
    --- PASS: TestParseProtocol/mixed_case:_Tcp (0.00s)
    --- PASS: TestParseProtocol/lower_case:_ssl (0.00s)
    --- PASS: TestParseProtocol/upper_case:_SSL (0.00s)
    --- PASS: TestParseProtocol/mixed_case:_SsL (0.00s)
    --- PASS: TestParseProtocol/invalid:_tls (0.00s)
    --- PASS: TestParseProtocol/empty (0.00s)
FAIL
FAIL    github.com/keep-network/keep-core/pkg/bitcoin/electrum  75.073s
FAIL
```

Reply:
I'm so sorry! I missed pushing the integration test updates. It should work fine now!

Reply:
Ok, It's working now.